### DISCUSSION TITLE
Licensing issues with Aphrodite Engine

### DISCUSSION THREAD

**Discussion by AlpinDale** [PygmalionAI/aphrodite-engine#37] [Announcements]
We have recently transitioned our license to [MIT](https://github.com/PygmalionAI/aphrodite-engine/pull/32) in response to concerns that the AGPL license might deter potential contributors and pose compatibility issues with non-AGPL licensed codebases. This decision was made following team discussions and direct exchanges with contributors. However, our initial choice of moving towards the less restrictive MIT license was not extensively deliberated upon. As a result, this change has limited our access to AGPL-licensed code from repositories such as [KoboldAI](https://github.com/henk717/KoboldAI) or [Text-Gen WebUI](https://github.com/oobabooga/text-generation-webui). 

The purpose of Aphrodite Engine is to offer a top-tier, innovative LLM inference engine that is available for *commercial* use, unlike [TGI](https://github.com/huggingface/text-generation-webui). The AGPL license would impose restrictions on any corporation utilizing proprietary software that integrates closely with this engine. As our forthcoming website is closed-source, we too would be adversely affected by the AGPL. 

Upon further internal deliberation, we concluded that while AGPL may be too restrictive, the MIT and Apache 2.0 licenses may be too *open*. This could potentially allow corporations to extract significant portions of Aphrodite code for use in their proprietary software, without any obligation to contribute back their modifications. A case in point is [Redis](https://redis.com/cloud-partners/aws/), where the BSD-3-Clause license permitted Amazon AWS to utilize, optimize, and enhance their codebase without any commitment to contribute upstream. We are keen to prevent such a scenario.

We anticipate that either LGPL or GPLv3 will fulfill our objectives, without discouraging corporate contributors. Given that we've already undergone one license change, we would prefer not to implement another so soon without an open discussion with all contributors to this repository. We invite you to share your thoughts on the adoption of GPLv3 or suggest an alternative license.  If you also have any questions, feel free to ask and I'll answer as best as I can.

**Comment by AlpinDale** [comment]
Tagging every contributor.
@TearGosling @StefanGliga @50h100a @g4rg

**Comment by g4rg** [comment]
Feel free to correct me if I'm wrong.

I don't think even GPL will help enforce the stated objective. It basically entitles the recepient of software to the source code and other things.
When the software, instead of being distributed to users, is utilized to serve users, modifications/improvements can remain private corporate intellectual property.
I believe this loophole to the spirit of GPL was the primary reason AGPL was created in the first place.

APGL affecting other software that integrates with the licensed work on an API level is new to me, and I see that as problematic if true.

**Reply by henk717** [reply]
I don't expect the latter to be true, but otherwise the projects developers should be able to state that they do not treat API usage as a derivative work that falls under the AGPL.

**Reply by AlpinDale** [reply]
I wonder if it's possible to state in the README, as an addition to the license, that API usage does not fall under AGPL license. If all contributors agree to this clause, I'm not sure what the problem would be.

**Reply by h4n1virus** [reply]
I am pretty sure it is not fine in just README and has to be in LICENSE itself for it to have any meaningful impact.

**Comment by StefanGliga** [comment]
Going for GPL or LGPL won't do the trick. The obligation to redistribute modified code triggers on redistribution, and for the purposes of those two, SaaS/a hosted instance does not count. In that case, they are functionally the same as BSD and MIT style licences.
AGPL, what we had before, is closest to what we want. There is no explicit clause in it forbidding hosting entirely(unlike HFOILv1.0 as used in HF TGI) and it is not 100% clear if AGPL does spread to other components of a bigger software system or not. I was under the impression it doesn't, but this thread prompted me to google it and it turns out there are lawyers claiming they could in good faith defend in court that it does.

I also researched ready-made alternative licences and I couldn't find anything better, all the other licences concerned with hosting I could find are even more restrictive.


It seem our 2 options are:

Option 1) Custom licence, written from scratch

Pros:
- We get maximum flexibility in specifying the freedom vs corporate friendliness tradeoff

Cons:
- We have to write an entire licence from scratch(tedious) and ideally run it past lawyers(expensive!!), or get lawyers to write one for us(very expensive!!!!)
- Licences not in wide use might scare off both FOSS enthusiasts and corporate users.


Option 2) AGPL with a carve out that software interacting with Aphrodite trough the API doesn't get infected with the licence. This might contradict a clause in the AGPL, as it explicitly mentions "scripts to control those activities" as being included(my reading is that it wouldn't but again it's ambigous and that's the problem here), so we would have to clarify that our carveout overrules the AGPL and be trough in listing everything in the carveout. GPL variants with carveouts aren't unprecedented, LGPL is the GPL with a carveout and GCC infamously ships with a carveout for code inserted as part of compilation

Pros:
- AGPL is an established licence, thoroughly covering any problem we might overlook

Cons:
- Just the 4 letter AGPL alone might scare of corporations anyway
- Carveouts might be seen as a custom licence, negating some benefits of using a widespread licence
- We might overlook something and unintentionally end up with a hostile licence anyway
- Writing a carveout is non trivial amount of work



We're stuck between a rock and a hardplace

**Comment by h4n1virus** [comment]
“The GNU Affero General Public License [...] requires the operator of a network server to provide the source code of the modified version running there to the users of that server. Therefore, public use of a modified version, on a publicly accessible server, gives the public access to the source code of the modified version"

"If you make access to modified software available to users over a network, you must make your source code available to those users."

"The license places restrictions on software used over a network which are extremely difficult for Google to comply with. Using AGPL software requires that anything it links to must also be licensed under the AGPL. Even if you think you aren’t linking to anything important, it still presents a huge risk to Google because of how integrated much of our code is. The risks heavily outweigh the benefits."

https://opensource.google/documentation/reference/using/agpl-policy

**Comment by AlpinDale** [comment]
After some deliberation with experts on this topic, we've come to the conclusion that we'll need to return to the AGPL license with no additional clauses if we are to continue using code from AGPL-licensed repositories such as @henk717 's [KoboldAI](http://github.com/henk717/KoboldAI). If we add a clause for API in the license, it would be nullified the moment a pure AGPL code is added to the repository. 

We want to keep collaborating with the Kobold community as close friends, so this isn't a risk we are willing to take. Those in favour of a return to AGPL, please leave an approval in the upcoming pull request. If you are against it, please let us know so we can figure out a solution.

**Reply by henk717** [reply]
I don't expect my own contributors to mind if the code is used by closed sourced frontends if that means we ger backend improvements because they have to contribute back the backend code.

I will have to ask them bur would not be surpriced if we adopt a sinilar clarification since it was our understanding of how it worked already when we wrote the code.

**Reply by AlpinDale** [reply]
I think it's a slippery slope. Adding an API exception clause would be void if we transfer an AGPL code to our repo, since the pure AGPL will override the modified one. And if you allow code transfers in your clause, then that could potentially be interpreted as Kobold allowing non-AGPL repositories to use their code.

**Reply by henk717** [reply]
For me its not a clause in general but a statement of our interpretation of the AGPL. Like our conversation yesterday this is a grey area if API's are considered linking. So we would be clarifying on our readme which usage we consider applies to it and which usage we do not consider effected by the AGPL.

For me personally that line is actually adding our code to a project, not using our webserver. So as long as lets say a closed source skyrim mod bundles the KoboldAI code its fine if all they do is connect to the API url.

So we would not be relicensing to a modified AGPL, instead we would make our own interpretation and intent clear to the users so that if this ever comes up in a court our interpretation can be used as a defense.

I still have to run it by our contributors but I am pretty sure everyone views AGPL that way on our side.

**Reply by henk717** [reply]
https://opensource.stackexchange.com/questions/4083/does-accessing-the-rest-interface-of-a-agpl-licensed-server-make-the-client-a-de

**Reply by StefanGliga** [reply]
My understanding of the entire issue is that software systems(like we wanted to unambiguously allow) would be considered more akin to compound works, not derivative works, and such wouldn't be subject to AGPL.
However I have found this reddit post https://www.reddit.com/r/programming/comments/m8amlr/comment/grhls7l/
That links back to HN thread where a user argues their lawyers warned them otherwise.
If this was my own project, I'd say to screw it and risk it, I don't think it would pass muster in court. But it's not solely mine so I wanted to cover all the bases.

**Comment by LostRuins** [comment]
> The purpose of Aphrodite Engine is to offer a top-tier, innovative LLM inference engine that is available for commercial use, unlike [TGI](https://github.com/huggingface/text-generation-webui). The AGPL license would impose restrictions on any corporation utilizing proprietary software that integrates closely with this engine. As our forthcoming website is closed-source, we too would be adversely affected by the AGPL.

Just wanna chime in that AGPL absolutely does allow for *commercial use* and indeed **anyone can take the code, modify it, and resell all or part it**, or run it as their own paid service, collecting all of the profits!

You could totally run a competitor offering a paid inference-as-a-service like chatgpt.

With just one main restriction - the final product would be open source. 

```
The AGPL License permits users of the licensed code to:

Use the code for commercial purposes: Like both GPL v2 and GPL v3, the AGPL imposes no conditions on using the code in software that’s sold commercially.
Modify the code: Users can change or rework the code, but if they distribute these changes/modifications in any public way (including over a server), they must release these updates in source code form under the AGPL license.
Distribute copies or modified/updated versions of the code: As long as these modifications are also released under the GNU AGPL, they can be distributed to others.
```

**Reply by AlpinDale** [reply]
Our main concern was that any frontend that interacts with the REST API provided by AGPL-licensed code could be construed as a derivative work. But if that's not the case, we'll happily switch back to AGPL. Henky linked [this article](https://opensource.stackexchange.com/questions/4083/does-accessing-the-rest-interface-of-a-agpl-licensed-server-make-the-client-a-de) which seems relevant. As such, we now have a PR at #38 to revert the license back to AGPL.
